samples/azure/azure-service-fabric-routing/Core_6/Shared/ReceiverSideDistribution/Pipeline/DistributeMessagesBasedOnPayload.cs

```diff
@@ -0,0 +1,71 @@
+using System;
+using System.Threading.Tasks;
+using NServiceBus;
+using NServiceBus.Pipeline;
+
+class DistributeMessagesBasedOnPayload :
+    IBehavior<IIncomingLogicalMessageContext, IIncomingLogicalMessageContext>
+{
+    string localPartitionKey;
+    Forwarder forwarder;
+    Func<object, string> mapper;
+
+    public DistributeMessagesBasedOnPayload(string localPartitionKey, Forwarder forwarder, Func<object, string> mapper)
+    {
+        this.localPartitionKey = localPartitionKey;
+        this.forwarder = forwarder;
+        this.mapper = mapper;
+    }
+
+    public Task Invoke(IIncomingLogicalMessageContext context, Func<IIncomingLogicalMessageContext, Task> next)
+    {
+        string messagePartitionKey;
+
+        var intent = GetMessageIntent(context);
+        var isSubscriptionMessage = intent == MessageIntentEnum.Subscribe || intent == MessageIntentEnum.Unsubscribe;
+        var isReply = intent == MessageIntentEnum.Reply;
+
+        if (isSubscriptionMessage || isReply)
+        {
+            return next(context);
+        }
+
+        // If the header is set we assume it's validity was checked by DistributeMessagesBasedOnHeader
+        if (context.MessageHeaders.ContainsKey(PartitionHeaders.PartitionKey))
+        {
+            return next(context);
+        }
+
+        messagePartitionKey = mapper(context.Message.Instance);
+
+        if (string.IsNullOrWhiteSpace(messagePartitionKey))
+        {
+            throw new PartitionMappingFailedException($"Could not map a partition key for message of type {context.Headers[Headers.EnclosedMessageTypes]}");
```

Comment:
Minor thingy, I personally don't like dynamic exception message for custom exceptions. Data in the exception should be set on the exception properties.